src/snowflake/cli/plugins/nativeapp/codegen/sandbox.py

```diff
@@ -0,0 +1,201 @@
+import os
+import shutil
+import subprocess
+import sys
+from enum import Enum
+from pathlib import Path
+from typing import Optional, Sequence, Union
+
+from click.exceptions import ClickException
+
+
+class SandboxExecutionError(ClickException):
+    """An error occurred while executing a python script."""
+
+    def __init__(self, error: str):
+        super().__init__(f"Failed to execute python script. {error}")
+
+
+def _get_active_venv_dir() -> Optional[str]:
+    return os.environ.get("VIRTUAL_ENV")
+
+
+def _get_active_conda_env() -> Optional[str]:
+    return os.environ.get("CONDA_DEFAULT_ENV")
+
+
+def _is_venv_active() -> bool:
+    return _get_active_venv_dir() is not None
+
+
+def _is_conda_active() -> bool:
+    return _get_active_conda_env() is not None
+
+
+def _is_ms_windows() -> bool:
+    return sys.platform == "win32"
+
+
+def _execute_python_interpreter(
+    python_executable: Optional[Union[str, Path, Sequence[str]]],
+    script_source: str,
+    cwd: Optional[Union[str, Path]],
+    timeout: Optional[int],
+) -> subprocess.CompletedProcess:
+    if not python_executable:
+        raise SandboxExecutionError("No python executable found")
+
+    if isinstance(python_executable, str) or isinstance(python_executable, Path):
+        args = [python_executable]
+    else:
+        args = [arg for arg in python_executable]
+    args.append("-")
+    return subprocess.run(
+        args,
+        capture_output=True,
+        text=True,
+        input=script_source,
+        timeout=timeout,
+        cwd=cwd,
+    )
+
+
+def _execute_in_venv(
+    script_source: str,
+    venv_path: Optional[Union[str, Path]] = None,
+    cwd: Optional[Union[str, Path]] = None,
+    timeout: Optional[int] = None,
+) -> subprocess.CompletedProcess:
+    resolved_venv_path = None
+    if venv_path is None:
+        active_venv_dir = _get_active_venv_dir()
+        if active_venv_dir is not None:
+            resolved_venv_path = Path(active_venv_dir).resolve()
+    elif isinstance(venv_path, str):
+        resolved_venv_path = Path(venv_path).resolve()
+    else:
+        resolved_venv_path = venv_path.resolve()
+
+    if resolved_venv_path is None:
+        raise SandboxExecutionError("No venv root found")
+
+    if not resolved_venv_path.is_dir():
+        raise SandboxExecutionError(
+            f"venv path should be an existing directory: {resolved_venv_path}"
+        )
+
+    # find the python interpreter for this environment. There is no need to activate environment prior to invoking the
+    # interpreter, as venv maintains the invariant that invoking any of the scripts will set up the virtual environment
+    # correctly. activation scripts are only used for convenience in interactive shells.
+    if _is_ms_windows():
+        python_executable = resolved_venv_path / "Scripts" / "python.exe"
```

Comment:
Can this `python.exe` for a venv be referring to an incompatible major version of python, such as `python 2.x`? Do we need to check for that? Specifying venv in my question as running the CLI itself requires 3.8 or higher.

Reply:
This is a utility function, it shouldn't be opinionated about python versions. If we want to check for supported versions, we should do it in the annotation processor preamble.

Reply:
That's fair, wherever the context screams "snowpark!" that's where we need it.